examples/multisig_plain/lib.rs

```diff
@@ -0,0 +1,231 @@
+
+        #[ink(message)]
+        fn change_requirement(&mut self, requirement: u32) {
+            self.ensure_from_wallet();
+            ensure_requirement(self.owners.len(), requirement);
+            self.required.set(requirement);
+        }
+
+        #[ink(message)]
+        fn submit_transaction(&mut self, transaction: Transaction) {
+            self.ensure_from_owner();
+            let id = self.transactions.put(transaction);
+            self.confirmation_count.insert(id, 0);
+            self.internal_confirm(id);
+        }
+
+        #[ink(message)]
+        fn cancel_transaction(&mut self, id: TransactionId) {
+            self.ensure_from_wallet();
+            self.take_transaction(id);
+        }
+
+        #[ink(message)]
+        fn confirm_transaction(&mut self, id: TransactionId) {
+            self.ensure_from_owner();
+            self.ensure_transaction_exists(id);
+            self.internal_confirm(id);
+        }
+
+        #[ink(message)]
+        fn revoke_confirmation(&mut self, id: TransactionId) {
+            self.ensure_from_owner();
+            if self
+                .confirmations
+                .remove(&(id, self.env().caller()))
+                .is_some()
+            {
+                self.confirmation_count
+                    .mutate_with(&id, |count| *count -= 1);
+            }
+        }
+
+        #[ink(message)]
+        fn execute_transaction(&mut self, id: TransactionId) -> Result<(), ()> {
+            self.ensure_confirmed(id);
+            let t = self.take_transaction(id).unwrap();
+            env::call::CallParams::<MyEnv, ()>::invoke(t.callee, t.selector.into())
+                .gas_limit(t.gas_limit)
+                .transferred_value(t.transferred_value)
+                .push_arg(&CallInput(&t.input))
+                .fire()
+                .map(|_| ())
+                .map_err(|_| ())
+        }
+
+        fn internal_confirm(&mut self, id: TransactionId) {
+            if self
+                .confirmations
+                .insert((id, self.env().caller()), ())
+                .is_none()
+            {
+                self.confirmation_count
+                    .mutate_with(&id, |count| *count += 1);
+            }
+        }
+
+        fn owner_index(&self, owner: &AccountId) -> u32 {
+            self.owners.iter().position(|x| *x == *owner).unwrap() as u32
+        }
+
+        fn take_transaction(&mut self, id: TransactionId) -> Option<Transaction> {
+            let transaction = self.transactions.take(id);
+            if transaction.is_some() {
+                self.clean_transaction_confirmations(id);
+            }
+            transaction
+        }
+
+        fn clean_owner_confirmations(&mut self, owner: &AccountId) {
+            for (id, _) in self.transactions.iter() {
+                if self.confirmations.remove(&(id, *owner)).is_some() {
+                    self.confirmation_count
+                        .mutate_with(&id, |count| *count -= 1);
```

Comment:
Couldn't we simply do this after clean-up of owners and set it directly to `0`? Would be more gas efficient.

Reply:
I see this being a performance hazard in case of many owners and transactions due to double-loop. We have to make sure that this is as efficient as possible and won't be a trap for gas costs.

Reply:
Are you sure you are commenting on the correct line? Is see no double loop here.